sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -153,6 +153,15 @@ object SetOperationPushDown extends Rule[LogicalPlan] with PredicateHelper {
         )
       )
 
+    // Push down limit into union
+    case Limit(exp, Union(left, right)) =>
+      Limit(exp,
+        Union(
+          CombineLimits(Limit(exp, left)),
+          CombineLimits(Limit(exp, right))
+        )
+      )
```

Comment:
Since we call `CombineLimits` here, we will not add extra `Limit` in the second iteration. Thus, I think it will cause the plan change. Thus, it will stop automatically, right?